### PR TITLE
Implement hasLanguage interop message for all enso objects

### PR DESCRIPTION
**Pull request by Akirathan**
Blocks #11468

### Pull Request Description

This PR implements the Truffle contract that all the Enso objects must export `hasLanguage` and `getLanguage` messages. By refactoring `EnsoObject` from an interface to an abstract class.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Run engine and stdlib benchmarks
- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
Quick note: I believe there are other ways to enforce the contract while keeping `EnsoObject` an interface.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java [hunk 1]
```diff
@@ -1,6 +1,27 @@
...
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.EnsoLanguage;
 
 /** All non-primitive Enso types extends from {@code EnsoObject}. */
-public interface EnsoObject extends TruffleObject {}
+@ExportLibrary(InteropLibrary.class)
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java:11]: I assume the goal is to provide default implementation of `hasLanguage` and `getLanguage` message for all Enso interop objects - e.g. those who implement `EnsoObject` interface.
- I would hope there is a better pattern to achieve this than converting to `abstract class`
- I don't like the `toDisplayString` to be defined, but unimplemented

Let's see if Truffle API offers some better alternative to achieve such a default common implementation of some `ExportLibrary` messages?

Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java:11]: > Let's see if Truffle API offers some better alternative to achieve such a default common implementation of some ExportLibrary messages?

I tried to leave `EnsoObject` as an interface, annotate it with `@ExportLibrary(InteropLibrary.class)`, and implement `hasLanguage` and `getLanguage` as default methods. But `@ExportMessage` on default method on interface does not work - the Truffle DSL processor fails with an error. Any ideas are welcome.

As for `toDisplayString` implementation - that is required by Truffle DSL, because if you export one of `getLanguage` or `hasLanguage` messages, then, `toDisplayString` is abstract and has to be implemented. However, I am still not sure what is the proper way to define it here, in `EnsoObject`.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java [hunk 2]
```diff
@@ -22,6 +22,6 @@ public Class<? extends TruffleLanguage<?>> getLanguage() {
 
   @ExportMessage
   public Object toDisplayString(boolean allowSideEffects) {
-    throw new UnsupportedOperationException("unimplemented");
+    return toString();
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java]: You cannot all `Object.toString()` in PE mode.

Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java]: Fixed by 7fa2022b8c6fd657057bc8da504e5c919df843ef

**Comment by Akirathan** [comment]
An approach with leaving `EnsoObject` as interfce, and defining a separate class `LanguageForEnsoObject` with  `@ExportLibrary(value = InteropLibrary.class, receiverType = EnsoObject.class)` does not work:
```diff
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
index ce60158116..eb65b641b2 100644
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
@@ -1,6 +1,15 @@
 package org.enso.interpreter.runtime.data;

 import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.DynamicDispatchLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;

 /** All non-primitive Enso types extends from {@code EnsoObject}. */
-public interface EnsoObject extends TruffleObject {}
+@ExportLibrary(DynamicDispatchLibrary.class)
+public interface EnsoObject extends TruffleObject {
+  @ExportMessage
+  Class<?> dispatch() {
+    return LanguageForEnsoObject.class;
+  }
+}
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/LanguageForEnsoObject.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/LanguageForEnsoObject.java
index cfd3eda4c0..6114f01e0e 100644
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/LanguageForEnsoObject.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/LanguageForEnsoObject.java
@@ -1,8 +1,10 @@
 package org.enso.interpreter.runtime.data;

+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.TruffleObject;
 import com.oracle.truffle.api.library.ExportLibrary;

-@ExportLibrary()
-public final class LanguageForEnsoObject {
-
+@ExportLibrary(value = InteropLibrary.class, receiverType = EnsoObject.class)
+public final class LanguageForEnsoObject implements TruffleObject {
+  // TODO: Export hasLanguage and getLanguage methods
 }
```

This patch fails to compile because Truffle DSL processor fails with errror: "@ExportLibrary is not supported for interfaces". And the receiver type must export `DynamicDispatchLibrary`.

TL;DR; `EnsoObject` cannot stay as an interface.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java:11]: I tried to use `@ExportLibrary(receiverType=EnsoObject.class)` but no luck. The best I could get with having an interface and also `Base` class is:
```diff
diff --git engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TFO.java engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TFO.java
new file mode 100644
index 0000000000..2545ad6dbe
--- /dev/null
+++ engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TFO.java
@@ -0,0 +1,15 @@
+package org.enso.interpreter.node.typecheck;
+
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.runtime.data.EnsoObject;
+
+@ExportLibrary(InteropLibrary.class)
+class TFO extends EnsoObject.Base implements EnsoObject {
+    @ExportMessage
+    @Override
+    public final String toDisplayString(boolean internal) {
+        return "TFO";
+    }
+}
diff --git engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
index ce60158116..f569c98256 100644
--- engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
+++ engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java
@@ -1,6 +1,24 @@
 package org.enso.interpreter.runtime.data;
 
+import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.EnsoLanguage;
 
 /** All non-primitive Enso types extends from {@code EnsoObject}. */
-public interface EnsoObject extends TruffleObject {}
+public interface EnsoObject extends TruffleObject {
+    @ExportLibrary(InteropLibrary.class)
+    public static abstract class Base implements EnsoObject {
+        @ExportMessage
+        public final boolean hasLanguage() {
+            return true;
+        }
+        @ExportMessage
+        public Class<? extends EnsoLanguage> getLanguage() {
+            return EnsoLanguage.class;
+        }
+        @ExportMessage
+        public abstract String toDisplayString(boolean internal);
+    }
+}
```

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/MetaObjectTest.java [hunk 3]
```diff
@@ -272,6 +274,33 @@ public void numbersAreEitherIntegerOrFloat() throws Exception {
     }
   }
 
+  /**
+   * Primitive values and exceptions currently don't have an associated language.
+   *
+   * <p>TODO[PM]: Will be implemented in https://github.com/enso-org/enso/pull/11468
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:280]: How do you want to associate a language with `java.lang.Long` or `java.lang.Boolean`? What language you'd like to associate with such values? JavaScript? Enso?

Language can only be associated with Enso objects - e.g. subtypes of `EnsoObject`!

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java:11]: According to @chumer, we should go with the `abstract class` - the `instanceof` checks will be even faster than with interfaces.

**Comment by JaroslavTulach** [comment]
> Quick note: I believe there are other ways to enforce the contract while keeping `EnsoObject` an interface.

I stay corrected. According to @chumer:

> yes interfaces are deliberately not supported as interface checks are expensive.
> Additionally they are a big pain to implement. (if not impossible)
> ...
> My recommendation is to make EnseBaseObject an abstract base class. You can still keep your interface

Let's turn `EnsoObject` into `abstract class` and gain the speed!

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java [hunk 4]
```diff
@@ -1,6 +1,30 @@
...
+  }
+
+  @ExportMessage
+  public Class<? extends TruffleLanguage<?>> getLanguage() {
+    return EnsoLanguage.class;
+  }
+
+  @ExportMessage
+  @TruffleBoundary
+  public Object toDisplayString(boolean allowSideEffects) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java]: This method can be `abstract`. I suggest to make it `abstract`. Definitely don't delegate to `toString()` - that's just a way for the guest language hacker to gain too much information about Enso interpreter internals.

**Review by JaroslavTulach — APPROVED** [review]
Turning `EnsoObject` into `abstract class` has a go!

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/MetaObjectTest.java (hunk 3 above)
Comment (Akirathan) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:280]: > How do you want to associate a language with java.lang.Long or java.lang.Boolean? What language you'd like to associate with such values? JavaScript? Enso?

Literal `42` in Enso is really not a literal, it is an atom of type `Integer`. So you can, e.g., call `42.up_to 80 . each ...`. The same is true for `True`, which is not the same as `java.lang.Boolean`. In the comment, I am talking about Enso "primitives" and not Java primitives.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java (hunk 4 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoObject.java]: Refactored to `abstract` in 46ebbd3e3fd5d1cf8e2e25be8076467d1d41a518 . Note that it must be `public`, otherwise Truffle DSL errors with something like:
```
[error] /home/pavel/dev/enso/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java:24:1:  error: Found invisible exported elements in super type 'EnsoObject': 
[error] public final class Warning extends EnsoObject {
[error]              ^     - org.enso.interpreter.runtime.data.EnsoObject.toDisplayString(boolean)
[error]   Increase their visibility to resolve this problem.
```

**Comment by Akirathan** [comment]
Benchmarks scheduled:
- Engine: https://github.com/enso-org/enso/actions/runs/11899805349
- Stdlib: https://github.com/enso-org/enso/actions/runs/11899829089

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@6dd7ac5</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Engine · enso-org/enso@6dd7ac5</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@6dd7ac5</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@6dd7ac5</div></blockquote>

**Comment by Akirathan** [comment]
stdlib benchmarks at https://github.com/enso-org/enso/actions/runs/11899829089 failed. Seems like a merge of newest develop will help here. Rescheduling benchmarks:
- engine: https://github.com/enso-org/enso/actions/runs/11912620823
- stdlib: https://github.com/enso-org/enso/actions/runs/11912626595

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@6dd7ac5</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@6dd7ac5</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@965f999</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Engine · enso-org/enso@965f999</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@965f999</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@965f999</div></blockquote>

**Comment by Akirathan** [comment]
No regressions in benchmarks, the PR is ready to merge.
